services/catalog/src/simcore_service_catalog/core/background_tasks.py

```diff
@@ -0,0 +1,177 @@
+import asyncio
+import logging
+from asyncio.futures import CancelledError
+from pprint import pformat
+from typing import Dict, List, Optional, Set, Tuple
+
+from aiopg.sa import Engine
+from aiopg.sa.connection import SAConnection
+from fastapi import FastAPI
+from pydantic import ValidationError
+from pydantic.types import PositiveInt
+
+from ..api.dependencies.director import get_director_session
+from ..db.repositories.groups import GroupsRepository
+from ..db.repositories.services import ServicesRepository
+from ..models.domain.service import (
+    ServiceAccessRightsAtDB,
+    ServiceDockerData,
+    ServiceMetaDataAtDB,
+)
+
+logger = logging.getLogger(__name__)
+
+ServiceKey = str
+ServiceVersion = str
+
+
+async def _list_registry_services(
+    app: FastAPI,
+) -> Dict[Tuple[ServiceKey, ServiceVersion], ServiceDockerData]:
+    client = get_director_session(app)
```

Comment:
use dependency injection instead (i.e. as you did in the handler functions) and add client as a parameter. That makes:
 - function more testable
 - dependencies more clear and less code
 - can reuse resources produced by dependencies when composed (e.g. share same client with other functions vs create/retrieve a client at every single call)

Reply:
I tried, but it did not work or I was too dumb to make it work...

Reply:
so after discussion with you, it seems dependency injection only works in handlers